src/services/registration/registration-router.ts

```diff
@@ -0,0 +1,81 @@
+import { Router } from "express";
+import { StatusCodes } from "http-status-codes";
+import {
+    RegistrationValidator,
+    RegistrationSchema,
+} from "./registration-schema";
+import { Database } from "../../database";
+import RoleChecker from "../../middleware/role-checker";
+import { Role } from "../auth/auth-models";
+
+const registrationRouter = Router();
+
+// A database upsert operation to save registration mid-progress
+registrationRouter.post(
+    "/save",
+    RoleChecker(["USER"]),
+    async (req, res, next) => {
+        try {
+            const registrationData = RegistrationValidator.parse(req.body);
+
```

Comment:
add a check in here to see if it's already submitted - if it is, then throw an error